### ISSUE TITLE
SampleProducer hangs on ProduceAsync()

### ISSUE THREAD

**Issue by naymore** [confluentinc/confluent-kafka-dotnet#190]
I'm making myself familiar with Kafka.
Because C# comes more natural to me I thought I might as well try this client.

The SampleProducer "hangs" on ProduceAsync() - it's never returning from that call. Not even after 30s.
Flush() is never called. Kafka is not showing anything. 

Any idea?

**Comment by mhowlett** [comment]
are you `await`ing the call? This may be issue #92.

**Comment by naymore** [comment]
In fact I was trying to send one single record.

```
using (Producer<string, string> myProducer = new Producer<string, string>(producerConfig, new StringSerializer(Encoding.UTF8), new StringSerializer(Encoding.UTF8)))
{
    Message<string, string> result = myProducer.ProduceAsync(_topicName, "Course-001", "My Message 1").Result;

    myProducer.Flush();
}
```

While debugging **I do not even hit the Flush()** method - not sure if #92 is evolving around that.
So yes, I believe there is something wrong with the async implementation.

**Comment by naymore** [comment]
Update.. ok, this is probably my fault - but maybe it helps ;-)

I tried to get some more information out of the producer.
So I added delegated for OnError et voilà.

> "jessie.local:9092/0: Failed to resolve 'jessie.local:9092': No such host is known. "

I learned upon first connection the producer learns about the whole cluster. I have a cluster with one single broker but still apparently the cluster info contains my "fake" DNS name - which of course I cannot resolve and hence the error.

**Comment by mhowlett** [comment]
good to know you ran into this - I want to fix it as a priority, but don't yet know how to do it without seriously affecting performance. See #180.

**Comment by treziac** [comment]
As a workaround for now, either call await Task.Yield() after ProduceAsync (this will put you back in the current thread) or do not use await (note that typically, you will not await ProduceAsync - this is perf killing as you won't have more than one request in flight. Depends of your usages)

**Comment by prabhuramsoft** [comment]
Hi
while using producer.ProduceAsync("....").Result, it **hangs** **Hangs always** :-( 

and when i try using Consumer, with code
 var config = new Dictionary<string, object>
            {
                {"group.id", "group"},
                {"bootstrap.servers", "hostname:portno"}
            };
var lstTopics = new List<string> { "Test_topic" };
 var cancelled = false;
  using (var consumer = new Consumer<Null, string>(config, null, new StringDeserializer(Encoding.UTF8)))
            {
                // Subscribe to the Kafka topic
                consumer.Subscribe(lstTopics);
                await Task.Yield();
                // Subscribe to the OnMessage event
                consumer.OnMessage += (obj, msg) =>
                {
                    Console.WriteLine($"Received: {msg.Value} by Consumer");
                };

                consumer.OnError += (obj, error) =>
                {
                    Console.WriteLine($"Error: {error}");
                    cancelled = true;
                };
                while (!cancelled)
                {
                    await Task.Run(() =>
                    {
                        consumer.Poll(TimeSpan.FromSeconds(1));
                    });
                }
            }
**seeing error like "hostname:portno/1003: Failed to resolve 'hostname:portno': No such host is known. "** 
Any help?

**Comment by treziac** [comment]
You are talking about consumer rather than producer.
Two things:
- you can't resolve the hostname/port. You have to give correct hostname/port for the library to be able to send data to kafka
- Your loop is not well designed. You are creating a new task every time you Poll. Instead, you should create a task to poll, like this:
  ```
  await Task.Run(() =>
  {
              while (!cancelled)
              {
                      consumer.Poll(TimeSpan.FromSeconds(1));
              }
  });
  ```
Refer to `example/AdvancedConsumer` for proper implementation

**Comment by prabhuramsoft** [comment]
**Producer logic below which hangs:**
  var config = new Dictionary<string, object> { { "bootstrap.servers", "hostname:Portno" } };
            var key = new byte[] { 1, 2, 3, 4 };
            var val = new byte[] { 5, 6, 7, 8 };
            var topic = "Test_topic";

            await Task.Run(() =>
            {
                // Create the producer
                using (var producer = new Producer<Null, string>(config, null, new StringSerializer(Encoding.UTF8)))
                {
                    // Send messages to the topic
                    for (var i = 0; i <= 10; i++)
                    {
                        var message = $"Event {i}";
                        var deliveryReport = producer.ProduceAsync(topic , null, message);
                        var res = deliveryReport.Result;      // **HANGS here**
                        Console.WriteLine(message);

                        producer.OnError += (_, error) =>
                        {
                            Console.WriteLine($"Producer Error: {error}");
                        };
                        producer.OnLog += (_, msg) =>
                        {
                            Console.WriteLine($"Producer Log: {msg.Message} Level: {msg.Level}");
                        };                    
                        }

                    // Calling flush method to flush out all messages
                    producer.Flush(100);
                }
            });

**Comment by mhowlett** [comment]
this was resolved and merged in #180
